File: libs/grpc/clients/watched.ts
import { ChannelCredentials, createChannel, createClient } from 'nice-grpc';

import { WatchedClient, WatchedDefinition } from '../generated/watched/watched.js';
import { PORTS } from '../servers/constants.js';
import { CLIENT_OPTIONS, createClientAddr, waitReady } from './helper.js';

export const createWatched = async (env: string): Promise<WatchedClient> => {
  const channel = createChannel(
    createClientAddr(env, 'watched', PORTS.WATCHED_SERVER_PORT),
    ChannelCredentials.createInsecure(),
    CLIENT_OPTIONS,
  );

  await waitReady(channel);

  const client = createClient(WatchedDefinition, channel);

  return client as any;
};

File: libs/grpc/clients/watched.go
package clients

import (
	"fmt"
	"log"

	"github.com/satont/tsuwari/libs/grpc/generated/watched"
	"github.com/satont/tsuwari/libs/grpc/servers"
	"google.golang.org/grpc"
	"google.golang.org/grpc/balancer/roundrobin"
	"google.golang.org/grpc/credentials/insecure"
)

func NewWatched(env string) watched.WatchedClient {
	serverAddress := createClientAddr(env, "watched", servers.WATCHED_SERVER_PORT)

	conn, err := grpc.Dial(
		serverAddress,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, roundrobin.Name)),
	)
	if err != nil {
		log.Fatalf("did not connect: %v", err)
	}
	c := watched.NewWatchedClient(conn)

	return c
}
